### PR TITLE
Add a new CR, a ComplianceCheckResult

### PR DESCRIPTION
**Pull request by jhrozek**
Adds a new CustomResource called ComplianceCheckResult that represents a result
of a check. A ComplianceCheckResult contains a unique ID of a check, a status and
a text description.

The parser was also changed a bit. Instead of creating the checks and the
remediations independently, let's change the parser to return a structure
holding a check and an optional remediation, which then allows for creating
the remediations in a single pass and as a result, the check can now act
as an owner of a remediation.

### CODE REVIEW

**Comment by jhrozek** [comment]
The test failure seems to be a flake:
```
time="2020-04-15T12:21:27Z" level=fatal msg="failed to create resource(s) in global resource manifest: Post https://api.ci-op-3lrng47n-d825e.origin-ci-int-aws.dev.rhcloud.com:6443/apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions: read tcp 10.128.167.87:60246->54.177.44.63:6443: read: connection reset by peer" 
```

**Comment by jhrozek** [comment]
The latest version renames the new CR as discussed on Slack.

File: cmd/remediation-aggregator/main.go [hunk 1]
```diff
@@ -189,36 +191,97 @@ func annotateParsedConfigMap(clientset *kubernetes.Clientset, cm *v1.ConfigMap)
...
 
-			fmt.Printf("Failed to create remediation: %v\n", err)
-			return err
+		if pr.Remediation == nil {
+			continue
 		}
+
+		remLabels := getRemediationLabels(scan)
+		if remLabels == nil {
+			return fmt.Errorf("cannot create remediation labels")
```
Comment (JAORMX) [inline cmd/remediation-aggregator/main.go:277]: This is gonna be really hard to figure out when it shows up in the logs. Let's instead return an error from the `getRemediatoinLabels` function, which will say exactly what happened.

File: cmd/resultserver/main.go [hunk 2]
```diff
@@ -123,7 +123,7 @@ func server(c *config) {
 		} else if encoding == "bzip2" {
 			extraExtension = "." + extraExtension
 		}
-		// TODO(jaosorior): Check that content-type is application/xml
+		// TODO(jaosorior): CheckResult that content-type is application/xml
```
Comment (JAORMX) [inline cmd/resultserver/main.go:126]: lol, let's revert back to the original message here.

File: deploy/operator.yaml [hunk 3]
```diff
@@ -33,9 +33,9 @@ spec:
             - name: OPENSCAP_IMAGE
               value: "quay.io/jhrozek/openscap-ocp:latest"
             - name: LOG_COLLECTOR_IMAGE
-              value: "quay.io/compliance-operator/resultscollector:latest"
+              value: "quay.io/compliance-operator/resultscollector"
             - name: RESULT_SERVER_IMAGE
-              value: "quay.io/compliance-operator/resultserver:latest"
+              value: "quay.io/compliance-operator/resultserver"
```
Comment (JAORMX) [inline deploy/operator.yaml:38]: Let's not change these.

File: pkg/apis/compliance/v1alpha1/complianceremediationcheck_types.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	// The check didn't yield a usable result
+	CheckResultNoResult ComplianceCheckStatus = ""
+)
+
+// ComplianceCheckResultSpec defines the desired state of ComplianceCheckResult
+type ComplianceCheckResultSpec struct {
+	// A unique identifier of a check
+	ID string `json:"id"`
+	// The result of a check
+	Result ComplianceCheckStatus `json:"result"`
```
Comment (JAORMX) [inline pkg/apis/compliance/v1alpha1/complianceremediationcheck_types.go:29]: Do we want to change this to be 'status' as well?

File: pkg/controller/compliancescan/compliancescan_controller.go [hunk 5]
```diff
@@ -220,7 +220,7 @@ func (r *ReconcileComplianceScan) phaseRunningHandler(instance *compv1alpha1.Com
 	}
 
 	if len(nodes.Items) == 0 {
-		log.Info("Warning: No eligible nodes. Check the nodeSelector.")
+		log.Info("Warning: No eligible nodes. CheckResult the nodeSelector.")
```
Comment (JAORMX) [inline pkg/controller/compliancescan/compliancescan_controller.go:223]: let's not change this log

File: pkg/controller/compliancesuite/compliancesuite_controller.go [hunk 6]
```diff
@@ -299,7 +299,7 @@ func (r *ReconcileComplianceSuite) reconcileRemediations(suite *compv1alpha1.Com
 		return reconcile.Result{}, err
 	}
 
-	// Check that all remediations have been applied yet. If not, cause an error
+	// CheckResult that all remediations have been applied yet. If not, cause an error
```
Comment (JAORMX) [inline pkg/controller/compliancesuite/compliancesuite_controller.go:302]: let's not change this comment

File: tests/e2e/helpers.go [hunk 7]
```diff
@@ -455,7 +485,7 @@ func waitForMachinePoolUpdate(t *testing.T, f *framework.Framework, name string,
 			pool.Status.UpdatedMachineCount, pool.Status.MachineCount,
 			pool.Status.UnavailableMachineCount)
 
-		// Check if the pool has finished updating yet. If the pool was paused, we just check that
+		// CheckResult if the pool has finished updating yet. If the pool was paused, we just check that
```
Comment (JAORMX) [inline tests/e2e/helpers.go:488]: let's not change this.

File: deploy/crds/compliance.openshift.io_v1alpha1_complianceremediationcheck_cr.yaml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+apiVersion: compliance.openshift.io/v1alpha1
+kind: ComplianceCheckResult
+metadata:
+  name: example-compliancecheckresult
+spec:
+  # Add fields here
+  size: 3
```
Comment (JAORMX) [inline deploy/crds/compliance.openshift.io_v1alpha1_complianceremediationcheck_cr.yaml:7]: Let's either add a real example here. Or just remove this file.

File: cmd/remediation-aggregator/main.go (hunk 1 above)
Comment (jhrozek) [reply cmd/remediation-aggregator/main.go:277]: fixed locally

File: cmd/resultserver/main.go (hunk 2 above)
Comment (jhrozek) [reply cmd/resultserver/main.go:126]: fixed locally

File: pkg/apis/compliance/v1alpha1/complianceremediationcheck_types.go (hunk 4 above)
Comment (jhrozek) [reply pkg/apis/compliance/v1alpha1/complianceremediationcheck_types.go:29]: Yes, thanks, fixed locally.

File: pkg/controller/compliancescan/compliancescan_controller.go (hunk 5 above)
Comment (jhrozek) [reply pkg/controller/compliancescan/compliancescan_controller.go:223]: fixed locally

File: pkg/controller/compliancesuite/compliancesuite_controller.go (hunk 6 above)
Comment (jhrozek) [reply pkg/controller/compliancesuite/compliancesuite_controller.go:302]: fixed locally

File: tests/e2e/helpers.go (hunk 7 above)
Comment (jhrozek) [reply tests/e2e/helpers.go:488]: Fixed locally.

File: deploy/operator.yaml (hunk 3 above)
Comment (jhrozek) [reply deploy/operator.yaml:38]: Fixed locally.

File: deploy/crds/compliance.openshift.io_v1alpha1_complianceremediationcheck_cr.yaml (hunk 8 above)
Comment (jhrozek) [reply deploy/crds/compliance.openshift.io_v1alpha1_complianceremediationcheck_cr.yaml:7]: fixed locally

**Comment by jhrozek** [comment]
Thank you for the review, here come new patches. I'm still waiting on the local e2e test to finish, but hopefully I haven't screwed up (famous last words..)

**Comment by jhrozek** [comment]
the last version had the deploy.yaml changes that removed `:latest` again, sorry. The latest version fixes that.

**Comment by jhrozek** [comment]
```
 level=fatal msg="Bootstrap failed to complete: waiting for Kubernetes API: context deadline exceeded" 
```

**Comment by jhrozek** [comment]
Sigh...

```
level=error msg="Attempted to gather ClusterOperator status after installation failure: listing ClusterOperator objects: the server could not find the requested resource (get clusteroperators.config.openshift.io)"
level=info msg="Pulling debug logs from the bootstrap machine"
level=info msg="Bootstrap gather logs captured here \"/tmp/artifacts/installer/log-bundle-20200416142938.tar.gz\""
level=fatal msg="Bootstrap failed to complete: waiting for Kubernetes API: context deadline exceeded"
---
```

**Comment by jhrozek** [comment]
/hold
there is a small bug, plus one more bug is coming up

**Comment by jhrozek** [comment]
btw the previous commits have not changed, to ease the review I added all the changes to a new commit

**Pull request merged**